Add metadata in imgpkg lock file

cari-lynn opened the pull request:
**When using the `--imgpkg-lock-output` flag, this PR adds the ` kbld.carvel.dev/metas` annotation.**
```
images:
- annotations:
    kbld.carvel.dev/metas: |
      - Tag: x.x.x
        Type: resolved
        URL: image-name:x.x.x
```

**When using a lock file as input to kbld, this PR also will preserve any ` kbld.carvel.dev/metas`**. It will append new metas to the list in the output and an imgpkg Lock file. This results in a history for example:
```
images:
- annotations:
    kbld.carvel.dev/metas: |
      - Tag: x.x.x
        Type: resolved
        URL: image-name:x.x.x
      - Type: preresolved
        URL: index.docker.io/library/..@sha256:...
```

**Other new behavior**:
- If a ` kbld.carvel.dev/metas` annotation is malformatted, kbld will error.
- the ` kbld.carvel.dev/metas` annotation must be in the format of several predefined objects, other unknown meta objects will be ignored.

cari-lynn committed:
Add `kbld.carvel.dev/metas` in imgpkg lock file

gcheadle-vmware committed:
Include image metas in output when provided by imgpkg lock file

Signed-off-by: Cari Dean <cdean@pivotal.io>

cari-lynn committed:
Update failing test expectations and fix typos

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

gcheadle-vmware committed:
Include 'Type: preresolved' as metadata in imgpkg lock file

- exclude metas in imgpkg lock when image is given in digest form
Signed-off-by: Cari Dean <cdean@pivotal.io>

gcheadle-vmware commented:
@cari-lynn Requesting review, was unable to select you as a GH reviewer.

cppforlife commented on pkg/kbld/cmd/resolve.go:
```diff
@@ -245,13 +246,38 @@ func (o *ResolveOptions) emitLockOutput(conf ctlconf.Conf, resolvedImages *Proce
 	}
 }
+
+func imgpkgLockAnnotations(i ProcessedImageItem) map[string]string {
```
lets place that on ResolveOptions so that it's not package global.

cppforlife commented on pkg/kbld/config/config.go:
```diff
@@ -67,6 +68,13 @@ type ImageOverride struct {
+type ImageMeta struct {
+	URL  string `json:"URL,omitempty" yaml:"URL,omitempty"`
+	Type string `json:"Type,omitempty" yaml:"Type,omitempty"`
+	Tag  string `json:"Tag,omitempty" yaml:"Tag,omitempty"`
```
i was under impression that some metas have other fields. i think we need to make this more generic.

cppforlife commented on pkg/kbld/config/config.go:
```diff
@@ -167,12 +175,15 @@ func NewConfigFromImagesLock(res ctlres.Resource) (Config, error) {
 
 	for _, image := range imagesLock.Images {
+		var iMetas []ImageMeta
+		yaml.Unmarshal([]byte(image.Annotations[ImagesLockKbldMetas]), &iMetas)
```
yaml.Unmarshal errors. what should we do with it?

cari-lynn replied:
If an imgpkg lock file cannot be marshalled into an ImageMeta struct, we should report an error that the lock file format was incorrect.

cppforlife commented on pkg/kbld/image/preresolved.go:
```diff
@@ -3,21 +3,34 @@
+		var givenMetas []Meta
+		for _, m := range i.metas {
+			givenMetas = append(givenMetas, PreresolvedImageSourceURL{Type: "preresolved", URL: m.URL, Tag: m.Tag})
+		}
```
not sure why we would be setting type to preresolved on given metas. if there are metas that indicate that it's git, it should remain as is. additionally we should append preresolved meta from line 35 to the results.

cppforlife commented on test/e2e/lock_output_test.go:
```diff
@@ -208,3 +227,78 @@ images:
+	if out != expectedOut {
+		t.Fatalf("Expected >>>%s<<< to match >>>%s<<<", out, expectedOut)
+	}
+
```
nit: remove newline

cari-lynn commented on pkg/kbld/image/preresolved.go:
```diff
@@ -3,21 +3,34 @@
 }
 
 func (i PreresolvedImage) URL() (string, []Meta, error) {
+	if len(i.metas) > 0 {
+		var givenMetas []Meta
+		for _, m := range i.metas {
+			givenMetas = append(givenMetas, PreresolvedImageSourceURL{Type: "preresolved", URL: m.URL, Tag: m.Tag})
+		}
+		return i.url, givenMetas, nil
+	}
```
Rename givenMetas -> imageMetas? Since 'given' doesn't map to anything related to this domain.

cari-lynn committed:
Change func to receiver

- return err from unmarshalling image lock

gcheadle-vmware committed:
Add preresolved meta to image metas from imgpkg lock

cari-lynn committed:
Add all image meta fields to ImageMeta struct
Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

cari-lynn committed:
Unmarshal ImageMetas from Lock file in config

- Use custom unmarshaler for image metas
- Move Metas from image package to config package
- Rename Metas to ImageMetas

cari-lynn committed:
Add license header :)

cari-lynn committed:
Reinitalize Metas for every item

cppforlife commented on pkg/kbld/image/preresolved.go:
```diff
@@ -3,21 +3,21 @@
-	return PreresolvedImage{url}
-}
+func (i PreresolvedImage) URL() (string, []ctlconf.Meta, error) {
+	imageMetas := append(i.metas, ctlconf.PreresolvedImageSourceURL{Type: "preresolved", URL: i.url})
```
why do we need to mutate PreresolvedImage?

pivotaljohn replied:
Making a copy before appending. 👍🏻

cppforlife commented on pkg/kbld/image/preresolved.go:
```diff
@@ -3,21 +3,21 @@
-	Type string // always set to 'preresolved'
-	URL  string
+func NewPreresolvedImage(url string, metas []ctlconf.Meta) PreresolvedImage {
+	return PreresolvedImage{url, metas}
```
we should make a copy of metas here to avoid accidentally mutating given array.

cppforlife commented on pkg/kbld/config/image_meta.go:
```diff
@@ -0,0 +1,97 @@
+			m.Metas = append(m.Metas, preres)
+		case yaml.Unmarshal(yamlItem, &tag) == nil && tag.Type == "tagged":
+			m.Metas = append(m.Metas, tag)
+		}
```
lets add 

```
default:
  // ignore unknown type
```

pivotaljohn replied:
See https://github.com/vmware-tanzu/carvel-kbld/pull/132#discussion_r707772960 for resolution.

cppforlife commented on pkg/kbld/config/image_meta.go:
```diff
@@ -0,0 +1,97 @@
+		yamlItem, _ := yaml.Marshal(&item)
+
+		switch {
+		case yaml.Unmarshal(yamlItem, &local) == nil && local.Type == "local":
```
should probably set these as consts since they are used elsewhere. (though ideally one wouldnt have to set Type manually at all)

pivotaljohn replied:
done.

cppforlife commented on pkg/kbld/config/image_meta.go:
```diff
@@ -0,0 +1,97 @@
+	meta()
+}
+
+type ImageMeta struct {
```
it seems that this type is used purely for unmarshaling. we should make it private to this package then.

pivotaljohn replied:
done.

cppforlife commented on pkg/kbld/config/image_meta.go:
```diff
@@ -0,0 +1,97 @@
+func (TaggedImageMeta) meta()           {}
+func (PreresolvedImageSourceURL) meta() {}
+
+func metasHistory(metas string) ([]Meta, error) {
```
metasHistory -> NewMetasFromString

pivotaljohn replied:
done.

cppforlife commented on pkg/kbld/config/config.go:
```diff
@@ -322,7 +329,8 @@ func UniqueImageOverrides(overrides []ImageOverride) []ImageOverride {
 		for _, addedOverride := range result {
-			if addedOverride == override {
+			// using DeepEqual instead of '==' since ImageOverride contains a slice
+			if reflect.DeepEqual(addedOverride, override) {
```
do we have any funky behaviour when we have same image multime times with different metas?

pivotaljohn replied:
Yeah, we want to look into what the possible "conflict" use cases there are... and how we want each to be resolved.  And if need be, make the appropriate changes.

pivotaljohn replied:
~~I think it looks like this:~~

- ~~no dups ==> no overrides removed~~
- ~~matching overrides (exact duplicates) ==> duplicates removed~~
- ~~equal "image[+imagerepo]":~~
  - ~~preresolved > !preresolved — respect presence of "preresolved" by not re-resolving~~
  - ~~!tagselection > tagselection — presume "newImage" without tagselection is more precise and therefore wins~~
  - ~~latest(tagselection) — for `semver` (current only strategy supported), whichever selects the largest/latest version value.~~
- ~~first override specified wins — if a set of overrides are not unique by this point, then take the first.~~

---
_edit: I misunderstood the intent, here; revised in next comment._

---
_edit: in fact, this is a bad idea, overall: it removes the control from consuming users who might want to truly override._

pivotaljohn replied:
oh wait... I think I misunderstood our intent, here.

- on reflection, I see that: _The order_ of overrides is the only thing that determines which one applies (i.e. first _always_ wins, regardless) — this is **not** changing
- ... rather ...: There are a couple of cases where additional meta are added (notably "preresolve"s). The concern here is making sure that we don't lose any meta on the way through de-duping. 
- also noted while reading the surrounding code more carefully: we don't actually call `UniqueImageOverrides()` when emitting an `imgpkg`-flavored lock file. I believe this is an oversight.

So, I'll review this flow with my revised understanding, today.

pivotaljohn replied:
Back to the original question:
> do we have any funky behaviour when we have same image multime times with different metas?

The answer is no... mostly because in all the code paths that lead to `UniqueImageOverrides()`, _right now_, none of the overrides have meta set.

This led me to step back and reflect on the overall workflow...

With this PR, we are propagating meta through `imgpkg` lock files.
However, we are _not_ doing so with plain `kbld` lock files.
Technically, this was not requested in either the issue or intended in this PR.

That said, if it's not unreasonable that users may which to use `kbld` lock files during development and then switch to `imgpkg` lock files when bundling up the config+images, then it would make sense to ensure that meta are carried through the `kbld` lock file flow.

I'm capturing this additional scope in a separate issue and it can be prioritized: #161.

cppforlife commented on pkg/kbld/config/config.go:
```diff
@@ -67,6 +68,7 @@ type ImageOverride struct {
 	NewImage     string                     `json:"newImage"`
 	Preresolved  bool                       `json:"preresolved,omitempty"`
 	TagSelection *versions.VersionSelection `json:"tagSelection,omitempty"`
+	ImageMetas   []Meta                     `json:",omitempty"`
```
seems like we should be specifying serialized name explicitly here. i do wonder if we should call it just metas though.

pivotaljohn replied:
Yeah, avoids accidental "breaking changes" if we rename the struct field.

I wish we had a more descriptive name than "metas:"

pivotaljohn replied:
added "metas" for now; will potentially rename in a separate commit.

cppforlife commented on pkg/kbld/cmd/resolve.go:
```diff
@@ -245,13 +246,31 @@ func (o *ResolveOptions) emitLockOutput(conf ctlconf.Conf, resolvedImages *Proce
+	anns := map[string]string{
+		ctlconf.ImagesLockKbldID: i.UnprocessedImageURL.URL,
+	}
+	imageMetas := i.Metas
```
not sure why we need local variable. use i.Metas directly?

pivotaljohn replied:
done.

cppforlife commented on pkg/kbld/cmd/resolve.go:
```diff
@@ -245,13 +246,31 @@ func (o *ResolveOptions) emitLockOutput(conf ctlconf.Conf, resolvedImages *Proce
+	}
+	imageMetas := i.Metas
+	if len(imageMetas) > 0 {
+		metaYaml, err := yaml.Marshal(imageMetas)
```
metaYaml -> metasYAML (or my favorite `bs`)

pivotaljohn replied:
that's just bullsh*t!

... done.

gcheadle-vmware commented:
This looks good to me. Great idea extracting out the meta interface into its own package, it really cleaned up nicely :)

pivotaljohn commented on pkg/kbld/cmd/resolve.go:
```diff
@@ -245,13 +246,31 @@ func (o *ResolveOptions) emitLockOutput(conf ctlconf.Conf, resolvedImages *Proce
 			iLock.Images = append(iLock.Images, lockconfig.ImageRef{
 				Image:       urlImagePair.Image.URL,
-				Annotations: map[string]string{ctlconf.ImagesLockKbldID: urlImagePair.UnprocessedImageURL.URL},
+				Annotations: anns,
```
**nit:** inline `anns`.

**rationale:**
- the variable `anns` doesn't seem to be doing much work (it's not reducing repeated invocations or saving memory, nor is it improving the readability)
- the function does not return an `error`
- the function describes well the returned result

pivotaljohn commented on pkg/kbld/image/resolved.go:
```diff
@@ -7,28 +7,21 @@ import (
 )
 
-// ResolvedImage respresents an image that will be resolved into url+digest
+// ResolvedImage represents an image that will be resolved into url+digest
```
❤️

pivotaljohn commented on pkg/kbld/image/tag_selected.go:
```diff
@@ -7,11 +7,12 @@ import (
 )
 
-// TagSelectedImage respresents an image that will be resolved into url+digest
+// TagSelectedImage represents an image that will be resolved into url+digest
```
😆

pivotaljohn commented on pkg/kbld/image/tagged.go:
```diff
@@ -9,25 +9,18 @@ import (
 )
 
-// TaggedImage respresents an image that will be tagged when its URL is requested
+// TaggedImage represents an image that will be tagged when its URL is requested
```
clearly copy pasta. Thanks for pulling these weeds.

pivotaljohn commented on pkg/kbld/config/image_meta.go:
```diff
@@ -0,0 +1,97 @@
+func (TaggedImageMeta) meta()           {}
+func (PreresolvedImageSourceURL) meta() {}
+
+func metasHistory(metas string) ([]Meta, error) {
```
```suggestion
func parseMetas(metas string) ([]Meta, error) {
```

It would be great if "Meta" had a better metaphor... and what I'm seeing here is an attempt to suggest one: something less reliable than "provenance" but tells a story from whence this image came: a history.

However, in _this_ context, this function is effectively taking the unmarshalling just one more step to include these `Meta` types. So, the intention of the function actually gets a bit blurred.

Another angle to "justify" this suggestion is that functions often _tend to_ work better as verb[-phrases] than nouns.

"Parse" rather than "unmarshall" because the function accepts an unstructured input and _produces_ a structured output, rather than stuffing a structure. 🤷🏻

pivotaljohn commented on pkg/kbld/config/image_meta.go:
```diff
@@ -0,0 +1,97 @@
+		yamlItem, _ := yaml.Marshal(&item)
+
+		switch {
+		case yaml.Unmarshal(yamlItem, &local) == nil && local.Type == "local":
```
Should we be deliberately ignoring unmarshallable meta?

I ask because when we ignore these errors some piece of this "history" is dropped without warning.
And while Carvel tools do not make direct use of this data, it is in effect an API; other tooling may well become dependent on the presence of it.

An alternative is to error out. If the user wants to ignore such an error, they can remove the meta annotations and try again.

pivotaljohn replied:
For now, dropping image meta (which plays the role of "resolution history"), is the way to go:
- meta are diagnostic information, not operational data
- reporting errors well (e.g. not flooding the output with warnings) adds complexity that's not worth the trade-off right now
- captured the possibility of doing more here: #160.

Added a comment noting that we're deliberately dropping meta (and why).

pivotaljohn reviewed (changes requested):
This work is _really_ close.

The overall approach looks really good to me.

Details in the comments, below.

pivotaljohn commented:
Reviewed w/ @garrettcheadle (offering to make changes if desired):

> i think that all suggestions but the last comment are simple changes that I see as improvements. ... If you could make those changes/renames, that would move this along.

👍🏻 

> In the case of "[deliberately ignoring unmarshallable data](https://github.com/vmware-tanzu/carvel-kbld/pull/132#discussion_r707772960)" ... we have another custom unmarshaller in kbld, maybe we should check if that unmarshaller has an error case for wisdom?

I'll look into that.

pivotaljohn commented:
In terms of prior art, I'm seeing this idiom which attempts to unmarshal data into a set of potential structs...

https://github.com/vmware-tanzu/carvel-kbld/blob/f919f7f78a/pkg/kbld/resources/path.go#L78-L91

By the end of the `switch`, if no suitable struct was found, that was reported as an error.

I think we should do a similar thing in the case of detecting which kind of `Meta` (if you can successfully unmarshall it, must be _that_ kind!)... but also either emit a warning or error out when it's unrecognized.

cppforlife reviewed (commented):
(looks like i had review that wasnt submitted. might be a bit stale.)

jtigger committed:
Fit and finish from review

- extract meta type names as constants
- fix slice append gotcha: when creating a new slice, copy from old
  first (i.e. copyAndAppendMeta())
- various renames and inlining to improve clarity

pivotaljohn commented:
Reviewed w/ @cppforlife over the phone.

- [x] `UniqueImageOverrides` is not needed when preparing to output `imgpkg`-flavored lock files. The other callers are in deprecated code. We'll remove this invocation from "resolve"
- [ ] to minimize the impact of this change on working ocde, we'll fix the compilation error, by writing a custom `Equals()` that ignores metas in the comparison. 👍🏻

jtigger committed:
Use a custom definition of equal ImageOverrides

- meta are descriptive, not identifying, so they have no bearing on
  whether two ImageOverrides are equal.

Given that all overrides are created brand fresh during a resolve for
the kbld lock file, there's no need to dedup them (removed call to
UniqueImageOverride()).
